### PR TITLE
[BIOMAGE-1284] Scaling the pipeline

### PR DESCRIPTION
**Pull request by kafkasl**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-1362

#### Link to staging deployment URL 

https://ui-grumpy-ochre-seahorse.scp-staging.biomage.net/
https://ui-martinfosco-api212-pipelin.scp-staging.biomage.net/

#### Links to any Pull Requests related to this

https://github.com/biomage-ltd/pipeline/pull/145
https://github.com/biomage-ltd/iac/pull/161

#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

Added new step to pipelines' skeletons so that the pipelines assign themselves to a given pod.

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [x] Unit tests written
- [ ] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [x] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "description": "Node.js API for the single-cell pipeline.",
   "main": "index.js",
   "scripts": {
-    "test": "jest --silent",
+    "test": "jest --silent --testTimeout=10000",
```
Comment (cosa65) [inline package.json:7]: This is to deal with the timeouts when testing locally?

File: src/api/general-services/pipeline-manage/index.js [hunk 2]
```diff
@@ -131,15 +131,20 @@ const createNewStateMachine = async (context, stateMachine, processName) => {
 };
 
 const executeStateMachine = async (stateMachineArn, execInput) => {
+  // when running in aws the step functions use a map step to retry the process
+  // of assigning the pipeline to an available pod
+  // map steps require an array as input so we declare one (it's value is not used)
+  const input = execInput || {};
```
Comment (cosa65) [inline src/api/general-services/pipeline-manage/index.js]: This `|| {}` can be defined as optional parameter in line 133, right?

File: src/api/general-services/pipeline-manage/skeletons/qc-pipeline-skeleton.js [hunk 3]
```diff
@@ -1,20 +1,10 @@
-const config = require('../../../../config');
+const { firstStep, buildInitialSteps } = require('./initialization');
 
-
-const qcPipelineSkeleton = {
-  Comment: `Pipeline for clusterEnv '${config.clusterEnv}'`,
-  StartAt: 'DeleteCompletedPipelineWorker',
+const getQcPipelineSkeleton = (clusterEnv) => ({
+  Comment: `Pipeline for clusterEnv '${clusterEnv}'`,
```
Comment (cosa65) [inline src/api/general-services/pipeline-manage/skeletons/qc-pipeline-skeleton.js]: I would change to `QC pipeline for clusterEnv '${clusterEnv}'`

File: src/config/default-config.js [hunk 4]
```diff
@@ -113,9 +113,9 @@ if (config.clusterEnv === 'development') {
     s3ForcePathStyle: true,
   });
 
-  // remove this line when the new gem2s-endpoint is merged and thus the production/pipeline.yaml is
-  // updated to new name pipeline-runner
-  config.pipelineInstanceConfigUrl = 'https://raw.githubusercontent.com/biomage-ltd/iac/master/releases/production/pipeline.yaml';
+  // TODO: remove this line when scaling work is merged and thus the production/pipeline.yaml is
```
Comment (cosa65) [inline src/config/default-config.js]: This should be done before closing this PR or later on? If it should be done later we should add a ticket so we don';t forget.

If it is before closing this PR we can use this comment as reminder

File: src/config/test-config.js [hunk 5]
```diff
@@ -16,7 +16,9 @@ module.exports = {
   workerNamespace: 'worker-test-namespace',
   pipelineNamespace: 'pipeline-test-namespace',
   workerInstanceConfigUrl: 'https://raw.githubusercontent.com/biomage-ltd/iac/master/releases/production/worker.yaml',
-  pipelineInstanceConfigUrl: 'https://raw.githubusercontent.com/biomage-ltd/iac/master/releases/production/pipeline.yaml',
+  // TODO: remove this line when scaling work is merged and thus the production/pipeline.yaml is
+  // updated to new name pipelineRunner
+  pipelineInstanceConfigUrl: 'https://raw.githubusercontent.com/biomage-ltd/iac/master/releases/staging-candidates/pipeline/refs-pull-145-merge.yaml',
```
Comment (cosa65) [inline src/config/test-config.js:22]: Reminder to remov this (or create ticekt)

File: src/api/general-services/pipeline-manage/constructors/assign-pod-to-pipeline.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    MaxConcurrency: 1,
+    // retry waits up to 226 seconds, fargate takes from 1 to 3 minutes to spawn a new pod
+    Retry: [{
+      ErrorEquals: ['NoPodsAvailable'],
+      IntervalSeconds: 2,
+      MaxAttempts: 11,
+      BackoffRate: 1.5,
+    }],
+    Iterator: {
+      StartAt: 'GetUnassignedPod',
```
Comment (cosa65) [inline src/api/general-services/pipeline-manage/constructors/assign-pod-to-pipeline.js:38]: In the case there is temporarily not an available and unassigned pod, could this step fail? In that case do we retry?

File: package.json (hunk 1 above)
Comment (kafkasl) [reply package.json:7]: yes, it fixes them for me and I assumed it wouldn't hurt other people.

File: src/api/general-services/pipeline-manage/index.js (hunk 2 above)
Comment (kafkasl) [reply src/api/general-services/pipeline-manage/index.js]: I didn't think about that, but then I get the style warning `Assignment to property of function parameter ` and I prefer to add the extra declaration than  to disable warnings

File: src/api/general-services/pipeline-manage/skeletons/qc-pipeline-skeleton.js (hunk 3 above)
Comment (kafkasl) [reply src/api/general-services/pipeline-manage/skeletons/qc-pipeline-skeleton.js]: done

File: src/config/default-config.js (hunk 4 above)
Comment (kafkasl) [reply src/config/default-config.js]: just before merging the PR

File: src/config/test-config.js (hunk 5 above)
Comment (kafkasl) [reply src/config/test-config.js:22]: same as before

File: src/api/general-services/pipeline-manage/index.js (hunk 2 above)
Comment (cosa65) [reply src/api/general-services/pipeline-manage/index.js]: I think the extra declaration is fine, it's just more readable and easier to see that this value is optional if its default value is defined in the paramters

**Review by cosa65 — APPROVED** [review]

File: src/api/general-services/pipeline-manage/index.js (hunk 2 above)
Comment (kafkasl) [reply src/api/general-services/pipeline-manage/index.js]: changed

**Pull request merged**